activation/wire/wire_v2.go

```diff
@@ -79,13 +171,62 @@ type MarriageCertificate struct {
 	Signature types.EdSignature
 }
 
+func (mc *MarriageCertificate) Root() []byte {
+	tree, err := merkle.NewTreeBuilder().
+		WithHashFunc(atxTreeHash).
+		Build()
+	if err != nil {
+		panic(err)
+	}
+	tree.AddLeaf(mc.ID.Bytes())
+	tree.AddLeaf(mc.Signature.Bytes())
+	return tree.Root()
+}
+
 // MerkleProofV2 proves membership of multiple challenges in a PoET membership merkle tree.
 type MerkleProofV2 struct {
 	// Nodes on path from leaf to root (not including leaf)
 	Nodes       []types.Hash32 `scale:"max=32"`
 	LeafIndices []uint64       `scale:"max=256"` // support merging up to 256 IDs
 }
 
+func (mp *MerkleProofV2) Root() []byte {
+	nodeTree, err := merkle.NewTreeBuilder().
+		WithHashFunc(atxTreeHash).
+		Build()
+	if err != nil {
+		panic(err)
+	}
+	for _, node := range mp.Nodes {
+		nodeTree.AddLeaf(node.Bytes())
+	}
+	for i := 0; i < 32-len(mp.Nodes); i++ {
+		nodeTree.AddLeaf(types.EmptyHash32.Bytes())
+	}
+	leafIndicesTree, err := merkle.NewTreeBuilder().
+		WithHashFunc(atxTreeHash).
+		Build()
+	if err != nil {
+		panic(err)
+	}
+	for _, index := range mp.LeafIndices {
+		leafIndicesTree.AddLeaf((*[8]byte)(unsafe.Pointer(&index))[:])
+	}
+	for i := 0; i < 256-len(mp.LeafIndices); i++ {
+		leafIndicesTree.AddLeaf([]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00})
+	}
+
+	merkleTree, err := merkle.NewTreeBuilder().
+		WithHashFunc(atxTreeHash).
+		Build()
+	if err != nil {
+		panic(err)
+	}
+	merkleTree.AddLeaf(nodeTree.Root())
+	merkleTree.AddLeaf(leafIndicesTree.Root())
+	return merkleTree.Root()
+}
```

Comment:
Why are there 3 trees needed instead of 1 for the `MerkleProofV2`?

Does it make sense to use merkle tree for this structure at all? Is there a usecase for proving a particular field of it? Maybe just encode with scale and hash?

Reply:
Changed to use the scale encoded merkleproof for the ID calculation.